src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,56 @@ public override bool Execute()
                 var filePath = Uri.Substring(FileUriProtocol.Length);
                 Log.LogMessage($"Copying '{filePath}' to '{DestinationPath}'");
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
-            else
-            {
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
+            Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
+
+            Random rng = new Random();
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
                     try
                     {
+                        var httpResponse = await httpClient.GetAsync(Uri);
+
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound)
+                        {
+                            if (!string.IsNullOrEmpty(PrivateUri))
+                            {
+                                httpResponse = await httpClient.GetAsync(PrivateUri);
+                            }
+
+                            // The Azure Storage REST API returns '400 - Bad Request' in some cases
+                            // where the resource is not found on the storage.
+                            // https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
+                            if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                                httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                            {
+                                return false;
```

Comment:
It doesn't look like it emits an error here either...

Reply:
Added a `throw FileNotFoundException` here. Going up in next push.